### PR TITLE
Napatech bypass v1.07

### PR DESCRIPTION
**Pull request by pjyoung1**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to redmine ticket:
[https://redmine.openinfosecfoundation.org/issues/2818](https://redmine.openinfosecfoundation.org/issues/2818)

Describe changes:
```
Added hardware based bypass support on Napatech SmartNICs
Added support for inline processing using Napatech SmartNICs
```
This PR Addresses issues identified in the review of [PR 4368 ](https://github.com/OISF/suricata/pull/4368)
This reflects changes since PR  [Napatech bypass v1.04](https://github.com/OISF/suricata/pull/4368)

### CODE REVIEW

File: src/source-napatech.c [hunk 1]
```diff
@@ -563,7 +1078,7 @@ TmEcode NapatechDecode(ThreadVars *tv, Packet *p, void *data, PacketQueue *pq,
             break;
         default:
             SCLogError(SC_ERR_DATALINK_UNIMPLEMENTED,
-                "Error: datalink type %" PRId32 " not yet supported in module NapatechDecode",
+                    "Error: datalink type %" PRId32 " not yet supported in module NapatechDecode",
```
Comment (jlucovsky) [inline src/source-napatech.c]: nit: `Error:` is redundant.

File: src/util-napatech.c [hunk 2]
```diff
@@ -23,28 +22,202 @@
...
+#include "util-byte.h"
 #include "threadvars.h"
 #include "tm-threads.h"
+#include "util-napatech.h"
+#include "source-napatech.h"
+
+#ifdef NAPATECH_ENABLE_BYPASS
+
+/* Counters to track the number of flows programmed on 
+the adapater.
```
Comment (jlucovsky) [inline src/util-napatech.c]: nit: change to ` * the adapter.`

File: src/util-napatech.c [hunk 3]
```diff
@@ -23,28 +22,202 @@
...
+    int status;
+
+    if ((status = NT_InfoOpen(&hInfo, "InfoStream")) != NT_SUCCESS) {
+        NAPATECH_ERROR(SC_ERR_NAPATECH_OPEN_FAILED, status);
+        exit(EXIT_FAILURE);
+    }
+
+    hInfoSys.cmd = NT_INFO_CMD_READ_SYSTEM;
+    if ((status = NT_InfoRead(hInfo, &hInfoSys)) != NT_SUCCESS) {
+        NAPATECH_ERROR(SC_ERR_NAPATECH_OPEN_FAILED, status);
```
Comment (jlucovsky) [inline src/util-napatech.c]: nit: extra leading space.

File: src/util-napatech.c [hunk 4]
```diff
@@ -23,28 +22,202 @@
...
+}
+
+/**
+ * \brief  Initializes the FlowStreams used to program flow data.
+ *
+ * Opens a FlowStream on each adapter present in the system.  This
+ * FlowStream is subsequently used to program the adapter with
+ * flows to bypass.
+ *
+ * \return count of the Napatech adapters present in the system.
```
Comment (jlucovsky) [inline src/util-napatech.c]: This is incorrect -- Returns bypass disabled (enabled) 0 (1).

File: src/util-napatech.c [hunk 5]
```diff
@@ -23,28 +22,202 @@
...
+
+/**
+ * \brief  Updates statistic counters for Napatech FlowStats
+ *
+ * \param tv     Thread variable to ThreadVars
+ * \param hInfo  Handle to the Napatech InfoStream.
+ * \param hstat_stream  Handle to the Napatech Statistics Stream.
+ * \param flow_counters The flow counters statistics to update.
+ *
+ * \param data      data pointer gets populated with
```
Comment (jlucovsky) [inline src/util-napatech.c]: Nit -- `data` isn't present
`clear_stats` isn't listed.

File: src/util-napatech.c [hunk 6]
```diff
@@ -61,18 +234,26 @@ enum CONFIG_SPECIFIER {
 
 #define MAX_HOSTBUFFERS 8
 
+/**
+ * \brief  Test to see if any of the configured streams are active
+ *
+ * \hInfo Handle to Napatech Info Stream.
+ * \hStatsStream Handle to Napatech Statistics stream
+ * \stream_config array of stream configuration structures
+ * \num_inst
```
Comment (jlucovsky) [inline src/util-napatech.c]: nit: Need `\param` for each parameter. 
`arg` isn't a param

File: src/util-napatech.c [hunk 7]
```diff
@@ -101,35 +279,55 @@ static uint16_t TestStreamConfig(
     return num_active;
 }
 
+/**
+ * \brief  Updates Napatech packet counters
+ *
+ * \tv Pointer to TheardVars structure
```
Comment (jlucovsky) [inline src/util-napatech.c]: nit: need `\param`

File: src/util-napatech.c [hunk 8]
```diff
@@ -101,35 +279,55 @@ static uint16_t TestStreamConfig(
...
+ * \num_streams the number of streams that are currently active
+ * \stream_config array of stream configuration structures
+ * \total_counters - cumulative count of all packets received.
+ * \dispatch_host, - Count of packets that were delivered to the host buffer
+ * \dispatch_drop - count of packets that were dropped as a result of a rule
+ * \dispatch_fwd - count of packets forwarded out the egress port as the result of a rule
+ * \is_inline - are we running in inline mode?
+ * \enable_stream_stats - are per thread/stream statistics enabled.
+ * \streamCounters - counters for each thread/stream configured.
+ *
```
Comment (jlucovsky) [inline src/util-napatech.c]: nit: missing return value?

File: src/util-napatech.c [hunk 9]
```diff
@@ -312,8 +726,15 @@ static void *NapatechStatsLoop(void *arg)
 #define MAX_STREAMS 256
 #define HB_HIGHWATER 2048 //1982
 
+/**
+ * \brief  Tests wheather a particular stream_id is actively registeed
+ *
+ * \stream_id - ID of the stream to look up
+ * \num_registered - The total number of registered streams
+ * \registered_streams - An array containing actively registered streams.
```
Comment (jlucovsky) [inline src/util-napatech.c]: nit: return value

**Comment by pjyoung1** [comment]
Addressed review comments in [Napatech_bypass_v1.09]( https://github.com/OISF/suricata/pull/4478)

**Pull request closed without merging**